Feature/feed parser

viacheslavbocharov opened the pull request:
- Mongo in Docker: docker-compose with a named volume and single-node replica set rs0; app connects via Prisma.
- Endpoint /feed: GET /feed?url=...&force=1
      Uses default URL when url is omitted.
      Cache in Mongo; with force=1 it parses upstream directly (skips cache).
      Parser: rss-parser + async-retry (retries with backoff; bails on 4xx).
- Routing: routes auto-registered via Fastify AutoLoad.
- Docs: Swagger as a plugin; UI available at /docs.
- Health checks: server — /health/server, database — /health/db.
- Logging: pino-pretty enabled in development for human-friendly logs.
- CI: GitHub Actions pipeline with Biome lint + build on PR.

viacheslavbocharov committed:
chore(logger): configure pino-pretty for dev environment

viacheslavbocharov committed:
chore(docker): configure MongoDB in docker-compose with env variables

viacheslavbocharov committed:
feat: add Prisma + Mongo integration and AutoLoad plugins

- Integrated Prisma ORM with MongoDB connection
- Added prisma plugin with Fastify type augmentation
- Connected plugins via AutoLoad for modular structure
- Prepared .env and schema.prisma for Mongo connection

viacheslavbocharov committed:
feat: add health endpoints for server and database

- Implemented /health/server endpoint with uptime and version info
- Implemented /health/db endpoint to check MongoDB connection via Prisma
- Verified endpoints return correct status responses

viacheslavbocharov committed:
feat: integrate RSS parser and implement /feed endpoint

- Connected rss-parser with async-retry for feed parsing
- Added /feed route with query params (url, force)
- Verified endpoint works and returns parsed data from https://hnrss.org/frontpage

viacheslavbocharov committed:
fix: RSS parser biome fix

viacheslavbocharov committed:
feat(parser): configure RSS parser and migrate routes to AutoLoad

- Implement GET /feed with url/force, retries, and ISO date normalization
- Move routes to src/routes and enable fastify-autoload

Gunnar97 commented on docker/mongo/mongod.conf:
```diff
@@ -0,0 +1,12 @@
+  replSetName: rs0
+
+security:
+  authorization: enabled
```
Круто, что сделал ключ. Если вдруг будут какие-то траблы, в принципе, можешь авторизацию задействовать для локальной разработки; она нам не сильно нужна.

viacheslavbocharov replied:
I will leave it as it is

Gunnar97 commented on src/app.ts:
```diff
@@ -21,8 +39,14 @@ async function buildApp(options: AppOptions = {}) {
+    fastify.log.info("Starting to load routes");
+    await fastify.register(AutoLoad, {
+      dir: join(__dirname, "routes"),
+      options,
```
Тут может быть проблема, так как твой автолоад будет смотреть на папку `routes` в корне. 

А так как у нас модули разбросаны по модулям, то тут необходимо это учесть.

viacheslavbocharov replied:
Fixed it.

Gunnar97 commented on src/config/schema.ts:
```diff
@@ -5,6 +5,7 @@ export const EnvSchema = {
   properties: {
     PORT: { type: "number" },
     HOST: { type: "string" },
+    DEFAULT_FEED_URL: { type: "string", nullable: true },
```
Не совсем понял, почему у тебя `default fit nullable true`, так как по заданию у нас дефолтная `urla` должна быть всегда.

viacheslavbocharov replied:
My bad. Fixed.

Gunnar97 commented on src/modules/feedParser/routes/feedParser.route.ts:
```diff
@@ -1,17 +1,165 @@
-}
+// import type { JsonSchemaToTsProvider } from '@fastify/type-provider-json-schema-to-ts';
+// import type { FastifyInstance } from 'fastify';
+// import {
```
Feed Parser должен находиться в модуле FeedParser.

viacheslavbocharov replied:
OK. Moved to module FeedParser

Gunnar97 commented on src/modules/feedParser/services/feedParserService.ts:
```diff
@@ -0,0 +1,99 @@
+
+// ⬇️ новый класс — чтобы в роуте легко различать 4xx
+export class Upstream4xxError extends Error {
+  constructor(
```
Непонятно, у нас в принципе весь сервис на функциональщине. Откуда у вас здесь появились классы?

viacheslavbocharov replied:
It was needed to rid of any. Rewrited with functional approach.

Gunnar97 commented on src/modules/feedParser/services/feedParserService.ts:
```diff
@@ -0,0 +1,99 @@
+        const s = extract4xx(err);
+        if (s) {
+          // ⬇️ bail: не ретраим 4xx
+          bail(new Upstream4xxError(`Upstream ${s}`, s));
```
Стараемся писать самостоятельно.

viacheslavbocharov replied:
OK

Gunnar97 commented on src/plugins/error-handler.plugin.ts:
```diff
@@ -0,0 +1,17 @@
+import fp from "fastify-plugin";
+
+export default fp(
```
Отпиши, для чего нам нужен этот плагин, пожалуйста.

viacheslavbocharov replied:
It’s a small global error-handler plugin to normalizes all uncaught errors into one response shape

Gunnar97 commented on src/plugins/error-handler.plugin.ts:
```diff
@@ -0,0 +1,17 @@
+      reply.status(statusCode).send({
+        statusCode,
+        code: err.code || "INTERNAL",
+        message: process.env.NODE_ENV === "development" ? err.message : "Internal error",
```
Здесь можем юзать наши конфиги, так как конфиг плагин у нас загружается раньше всех других плагинов.

viacheslavbocharov replied:
Good point to switched the error handler to use our config

Gunnar97 commented on src/plugins/prisma.plugin.ts:
```diff
@@ -0,0 +1,14 @@
+export default fp(
+  async (fastify) => {
+    const prisma = new PrismaClient();
+    await prisma.$connect();
```
Не вижу `try/catch`.

viacheslavbocharov replied:
Fixed

Gunnar97 commented on src/routes/feed.route.ts:
```diff
@@ -0,0 +1,71 @@
+import { parseFeed, Upstream4xxError } from "../modules/feedParser/services/feedParserService";
+import { getFeedCacheByUrl, getLatestItems } from "../services/dbService";
+
+export default fp(async function routes(app: FastifyInstance) {
```
Я так понял, это еще не готово. 

Поэтому еще раз напоминаю, что в роуте у нас находится только логика хендлинга запроса и ответа — удачного либо неудачного. 

Вся остальная бизнес-логика с query-параметрами и тому подобное мы помещаем в сервисы и работаем уже с сервисами.

viacheslavbocharov replied:
Deleted this. Mover route to the FeedParser module.

Gunnar97 commented on src/routes/health.db.route.ts:
```diff
@@ -0,0 +1,11 @@
+export default async function (app: FastifyInstance) {
+  app.get("/health/db", async (_req, reply) => {
+    try {
+      await app.prisma.feedCache.findFirst({ take: 1 });
```
В принципе, health check можно поместить в один файл roads, типа health roads, и на том все.

viacheslavbocharov replied:
Combine in one health.route

viacheslavbocharov commented:
Did all fixes, ready for review

viacheslavbocharov committed:
fix(feed-parser): apply review fixes and stabilize

Env: require DEFAULT_FEED_URL (URI) and add NODE_ENV; use app.config instead of process.env.

Error handler: unified response; map 4xx/ENOTFOUND/timeouts; hide stack in prod.

Parser: drop class; use functional makeUpstream4xx + isUpstream4xx.

AutoLoad: load *.route.* from src/routes and src/modules only.

Prisma plugin: add try/catch on $connect, logging, graceful onClose.

Health: merge /health/server and /health/db into one routes file.

Types/lint: remove non-null assertions, no any, small cleanups.

viacheslavbocharov committed:
fix: biom linting

viacheslavbocharov committed:
fix: move logic to getFeedResponse.service.ts,  add error handlet to the empty catch, used @fastify/sensible

viacheslavbocharov committed:
fix: /feed typing & logging, extract service, DB health ping

/feed: schema-typed req.query + req.log.error; delegate to getFeedResponse(app, url, force).

service: add getFeedResponse returning FeedResponse (cache + parse inside).

health: DB check via prisma.$runCommandRaw({ ping: 1 }), warn log, return 503 on failure.

schema/types: export FeedResponse and align response typing.

viacheslavbocharov committed:
Merge pull request #2 from viacheslavbocharov/fix/feed-parser

Fix/feed parser
